src/modm/platform/clock/sam_pmc/clockgen_impl.hpp.in

```diff
@@ -11,34 +11,89 @@
 
 namespace modm::platform {
 
-template <MasterClkSource src, MasterClkPrescaler pres>
+template <MainInternalFreq freq>
+void ClockGen::enableMainInternal() {
+	// Enable Main RC Oscillator and wait for it to stabilize
+	PMC->CKGR_MOR |= CKGR_MOR_KEY_PASSWD | CKGR_MOR_MOSCRCEN;
+	while (!(PMC->PMC_SR & PMC_SR_MOSCRCS));
+
+	// Change Main RC Oscillator frequency
+	PMC->CKGR_MOR =
+		(PMC->CKGR_MOR & ~(CKGR_MOR_MOSCRCF_Msk)) |
+		CKGR_MOR_MOSCRCF((uint32_t)freq) |
+		CKGR_MOR_KEY_PASSWD;
+	while (!(PMC->PMC_SR & PMC_SR_MOSCRCS));
+}
+
+template <MainClockSource src>
+void ClockGen::selectMainClockSource() {
+	if (src == MainClockSource::Internal) {
+		PMC->CKGR_MOR =
+			(PMC->CKGR_MOR & ~CKGR_MOR_MOSCSEL) |
+			CKGR_MOR_KEY_PASSWD;
+	} else {  // MainClockSource::External
+		PMC->CKGR_MOR |= CKGR_MOR_MOSCSEL | CKGR_MOR_KEY_PASSWD;
+	}
+}
+
+template <MasterClkSource src, MasterClkPrescaler pres, MasterClkDivider divider>
 void ClockGen::selectMasterClk() {
 	// Datasheet says when selecting PLL as source, write PRES first, otherwise
 	// write CSS first.
-	if(src == MasterClkSource::PLLA_CLK || src == MasterClkSource::PLLB_CLK) {
+%% if target["family"] in ["g"]
+	static_assert(divider == MasterClkDivider::Div1, "Divider for master clock is not supported");
+	constexpr bool isPll = (src == MasterClkSource::PLLA_CLK
+		|| src == MasterClkSource::PLLB_CLK);
+%% elif target["family"] in ["v"]
+	constexpr bool isPll = (src == MasterClkSource::PLLA_CLK
+		|| src == MasterClkSource::UPLL_CLK);
+%% endif
+
+	if constexpr (isPll) {
 		PMC->PMC_MCKR = (PMC->PMC_MCKR & ~PMC_MCKR_PRES_Msk) | PMC_MCKR_PRES((uint32_t)pres);
 		while(!(PMC->PMC_SR & PMC_SR_MCKRDY)) {}
+%% if target["family"] in ["v"]
+		PMC->PMC_MCKR = (PMC->PMC_MCKR & ~PMC_MCKR_MDIV_Msk) | PMC_MCKR_MDIV(uint32_t(divider));
+		while(!(PMC->PMC_SR & PMC_SR_MCKRDY)) {}
+%% endif
 		PMC->PMC_MCKR = (PMC->PMC_MCKR & ~PMC_MCKR_CSS_Msk) | PMC_MCKR_CSS((uint32_t)src);
 		while(!(PMC->PMC_SR & PMC_SR_MCKRDY)) {}
 	} else {
 		PMC->PMC_MCKR = (PMC->PMC_MCKR & ~PMC_MCKR_CSS_Msk) | PMC_MCKR_CSS((uint32_t)src);
 		while(!(PMC->PMC_SR & PMC_SR_MCKRDY)) {}
 		PMC->PMC_MCKR = (PMC->PMC_MCKR & ~PMC_MCKR_PRES_Msk) | PMC_MCKR_PRES((uint32_t)pres);
 		while(!(PMC->PMC_SR & PMC_SR_MCKRDY)) {}
+%% if target["family"] in ["v"]
+		PMC->PMC_MCKR = (PMC->PMC_MCKR & ~PMC_MCKR_MDIV_Msk) | PMC_MCKR_MDIV(uint32_t(divider));
+		while(!(PMC->PMC_SR & PMC_SR_MCKRDY)) {}
+%% endif
 	}
 }
 
-template<uint32_t multiplier>
+template<uint32_t multiplier, uint32_t divider>
 void ClockGen::enablePllA(uint32_t wait_cycles) {
+%% if target["family"] in ["g"]
     static_assert(multiplier > 8 && multiplier < 7502, "Valid PLL MUL range is 9-7501");
```

Comment:
```suggestion
    static_assert(9 <= multiplier && multiplier <= 7501, "Valid PLL MUL range is 9-7501");
```
Visually easier to grok due to similarity with `9 ≤ multiplier ≤ 7501` and `[9, 7501]` intervall notation style.